File: big/src/big.go
/* Compile and run under Linux and macOS with "go run big.go"

This implementation uses synchronous channels: each worker is a goroutine that
has channel ping for incoming requests and for outgoing requests as well as a
channel pong for outgoing replies and incoming replies.

Once a worker is done sending the specified number of requests, it notifies the
supervisor that it is done but keeps accepting requests. The supervisor waits
to be notified from all workers and then terminates as well.
*/
package main
import "fmt"

const Workers = 100
const Neighbourhoods = 10
const Rounds = 10000

func worker(id int, pings [Workers]chan int, pongs [Workers]chan int,
	done chan int) {
	rand := 12345 + id        // every worker has unique seed
	recipient := func() int { // generates random worker different from id
		for {
			rand = (32236 * rand) % 65521 // 2^16 - 15
			if rand % Workers != id {
				return rand % Workers
			}
		}
	}
	pingpong := 0
	r := 0 // number of rounds
	for r < Rounds {
		select {
		case pings[recipient()] <- id: // send ping to random neighbour
			recid := <-pongs[id]  // wait for pong to be returned
			pingpong++; r++
			if recid != id {panic("received id incorrect")}
		case other := <-pings[id]: // receive ping, send received id back again
			pongs[other] <- other
			pingpong--
		}
	}
	done <- pingpong // done with sending pings; still accepting incoming pings
	for {
		other := <-pings[id]
		pongs[other] <- other
	}
}
func supervisor() {
	done := make(chan int)
	for n := 0; n < Neighbourhoods; n++ {
		var pings, pongs [Workers]chan int
		for w := 0; w < Workers; w++ {
			pings[w], pongs[w] = make(chan int), make(chan int)
		}
		for w := 0; w < Workers; w++ {
			go worker(w, pings, pongs, done)
		}
	}
	min, max := 0, 0
	for w := 0; w < Workers * Neighbourhoods; w++ {
		pingpong := <-done
		if pingpong < min {
			min = pingpong
		} else if pingpong > max {
			max = pingpong
		}
	}
	fmt.Println("min, max:", min, max)
}
func main() {
	supervisor()
}

File: big/src/big.java
/* Compile under Linux and macOS with "javac big.java"; run with
"java Supervisor"

This implementation of the Big benchmark uses Java monitors. As calls in Java
are closed, i.e. the lock to a monitor is preserved while execution leaves the
monitor, a naive implementation with worker A calling ping of worker B who calls
back pong of worker A is prone to deadlock (nested monitor calls) if worker B
also pings worker A. Each ping-pong interaction involves exactly two workers.
The solution here is deadlock-free: the workers are ordered according to their
id and the worker initiating the ping-pong interaction first obtains a lock to
itself or to the recpient, whichever has the smaller id.

Once a worker thread is done with the specified number of requests, it
terminates. The supervisor waits to be notified from all workers and then
terminates as well.
*/
class Worker extends Thread {
  int id;
  Worker[] neighbours;
  Supervisor sup;
  int rand;
  int pingpong = 0;
  private int recipient() { // returns random neighbour different from id
    while (true) {
      rand = (32236 * rand) % 65521; // 2^16 - 15
      int result = rand % Supervisor.Workers;
      if (result != id) return result;
    }
  }
  Worker (int id, Worker[] neighbours, Supervisor sup) {
    this.id = id; this.neighbours = neighbours; this.sup = sup;
    rand = 12345 + id; // every worker has unique seed
  }
  synchronized void ping(Worker w, int id) {
    w.pong(id);
  }
  synchronized void pong(int id) {
    assert this.id == id : "received id incorrect";
    pingpong -= 1;
  }
  public void run() {
    for (int r = 0; r < Supervisor.Rounds; r++) {
      pingpong +=1;
      int n = recipient();
      if (n < id) { // first neighbours[n] is locked, then this is locked
        neighbours[n].ping(this, id);
      } else { // first this is locked, then neighbours[n] is locked
        synchronized(this) {neighbours[n].ping(this, id);}
      }
    }
    sup.done(pingpong);
  }
}
class Supervisor extends Thread {
  static final int Workers = 100;
  static final int Neighbourhoods = 10;
  static final int Rounds = 10000;
  int min, max, replies;
  synchronized void done(int pingpong) {
    if (pingpong < min) min = pingpong;
    else if (pingpong > max) max = pingpong;
    replies += 1;
    if (replies == Workers * Neighbourhoods) notify();
  }
  synchronized public void run() {
    for (int n = 0; n < Neighbourhoods; n++) {
      Worker[] neighbourhood = new Worker[Workers];
      for (int w = 0; w < Workers; w++) {
        neighbourhood[w] = new Worker(w, neighbourhood, this);
      }
      for (int w = 0; w < Workers; w++) {
        neighbourhood[w].start();
      }
    }
    try {wait();} catch (InterruptedException e) {}
    System.out.println("min, max: " + min + ", " + max);
  }
  public static void main(String args[]) {
    new Supervisor().start();
  }
}
